WIP - async connection

bdonlan opened the pull request:
This is an early draft of an asynchronous websocket connection to discord, (obviously) not for merge yet, but I'd like to hear comments on the approach as it's been implemented so far.

Currently supported:

* Receiving messages
* (probably) resume and reconnect (but it needs more testing)

TODO:

* Reimplement the old Connection interface on top of this
* voice support in general

bdonlan committed:
WIP - async connection

This is an early draft of an asynchronous websocket connection to discord.

Currently supported:

* Receiving messages
* (probably) resume and reconnect (but it needs more testing)

TODO:

* Reimplement the old Connection interface on top of this
* voice support in general

tinaun commented:
this is thorough! nice. i'll comment what i can

tinaun commented on Cargo.toml:
```diff
@@ -39,6 +50,6 @@ default-features = false
 optional = true
 
 [dependencies.multipart]
```
the good part about depending on reqwest is that we don't need this or the tls crates at all, both implicitly do the right thing

bdonlan replied:
Sure, I didn't want to have to mess with the synchronous code too much while focusing on async is all. That's also why it's on hyper 0.10, as 0.11 introduced a bunch of api changes apparently while 0.10 was compatible with everything else I needed while still mostly compatible with the existing `Discord` impls.

tinaun commented on src/async/mod.rs:
```diff
@@ -0,0 +1,127 @@
+    use serde_json;
+    use error::Error;
+    // this trait matches websocket connections wrapped by the serializer
+    pub trait Serializedish : Stream<Item=GatewayEvent, Error=Error>
```
i don't like the names but all these are great, nice structure

bdonlan replied:
Placeholder names :)

tinaun commented on src/async/pingfilter.rs:
```diff
@@ -0,0 +1,109 @@
+use super::imports::*;
+use websocket::{OwnedMessage, WebSocketError};
+
```
do we need this? its nice but i don't know if we ever had to handle this in the sync version

bdonlan replied:
Not sure. I included it mostly for completeness. If discord never actually sends WS-level pings it might not matter.

tinaun commented on src/async_connection.rs:
```diff
@@ -0,0 +1,543 @@
+use std::io;
```
is this another (older? version) of the system in `src/async`? it seems weird and self contained

bdonlan replied:
Yeah, this was a prototype I wrote while trying to figure out how tokio worked; I ended up splitting it out as it got more complex and hadn't taken the time to clean up the older cruft in the branch yet.

tinaun commented on src/connection.rs:
```diff
@@ -1,47 +1,40 @@
 }
 
 /// Websocket connection to the Discord servers.
 pub struct Connection {
```
in my work, i pushed the core to its own separate thread and pushed futures into it. i used regular std mpsc queues to handle constructing futures and allowed a cleaner seperation of sync and async code. but i don't know if that way of structuring things would even work all that well

bdonlan replied:
Yeah, I think the way to go with implementing the sync API is to put the core off on a thread. Some of the futures we work with (and the core/handle itself) aren't Send so it'd be a pretty big compatibility break not to spin it off.

bdonlan committed:
make resume work

bdonlan committed:
rm old async_connection

bdonlan committed:
Restore basic synchronous connection functionality

bdonlan committed:
Port back non-voice commands

bdonlan commented:
With these changes the synchronous Connection API is back, implemented on top of the async backend. Next up is getting voice working. I suspect it'll be easiest to just reimplement voice as an async backend too rather than trying to hack in the new websocket sync interface (given that it doesn't support `.split()` on SSL connections...), so I'll be working on that next, then go back and tidy everything up.

jD91mZM2 commented:
So is this thing using tokio asyncs or threads? Are mutexes required to access variables? Sometimes? Always?

tinaun commented:
@legolord208 both, using a tokio event loop running the async connection in a seperate thread, connected by two channels. (note that the old connection logic also required a seperate thread for the heartbeat) ideally both the voice and event connections would run in the same event loop, so that a new voice connection doesn't have to take `&mut` and would run in the same thread

jD91mZM2 commented:
:+1:

bdonlan commented:
This is using tokio asyncs; currently the only threads used are 1) the tokio timer thread and 2) a thread spawned to obtain the gateway URL (this can be fixed, but I haven't gotten around to it).

Mutexes are currently being used in a bunch of places, mostly just because `BoxFuture` requires `Send`. I plan to explore refactoring to use a non-`Send` boxed future type instead (we're already non-`Send` because we carry around a `Handle`) later.

For voice, I'd like to support either cross-thread or same-thread voice connections. This would allow busy bots to farm voice processing to multiple threads, while allowing simpler bots to use a single thread for everything. Still working out how this will look API-wise, but it ought to be possible to avoid mutexes for this in the single-thread case, I think.

tinaun commented:
any updates?

bdonlan commented:
Haven't had much time to work on this recently. Might be able to put more time into it next week.

bdonlan commented:
Closing for now as I probably won't have time to work on this soon.
